cmd/osbuild-jobsite-builder/main.go

```diff
@@ -119,45 +118,42 @@ func (builder *Builder) GetState() State {
 	return builder.State
 }
 
-func (builder *Builder) GuardState(stateWanted State) {
+func (builder *Builder) GuardState(stateWanted State) error {
 	if stateCurrent := builder.GetState(); stateWanted != stateCurrent {
-		logrus.Fatalf("Builder.GuardState: Requested guard for %d but we're in %d. Exit", stateWanted, stateCurrent)
+		return fmt.Errorf("Builder.GuardState: Requested guard for %d but we're in %d. Exit", stateWanted, stateCurrent)
 	}
+
+	return nil
 }
 
-func (builder *Builder) RegisterHandler(h Handler) http.HandlerFunc {
+func (builder *Builder) RegisterHandler(s State, m string, h Handler) http.HandlerFunc {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if err := builder.GuardState(s); err != nil {
+			w.WriteHeader(http.StatusConflict)
+			logrus.Fatal("Received request in invalid state. Exiting")
+		}
+
+		if r.Method != m {
+			w.WriteHeader(http.StatusBadRequest)
+			logrus.Fatal("Received invalid request method. Exiting")
+		}
+
 		if err := h(w, r); err != nil {
 			logrus.Fatal(err)
 		}
 	})
 }
 
-func (builder *Builder) HandleClaim(w http.ResponseWriter, r *http.Request) error {
-	builder.GuardState(StateClaim)
-
-	if r.Method != "POST" {
-		logrus.WithFields(
-			logrus.Fields{"method": r.Method},
-		).Fatal("Builder.HandleClaim: unexpected request method")
-	}
-
-	fmt.Fprintf(w, "%s", "done")
-
+func (b *Builder) HandleClaim(w http.ResponseWriter, r *http.Request) error {
 	logrus.Info("Builder.HandleClaim: Done")
 
-	builder.SetState(StateProvision)
+	w.WriteHeader(http.StatusOK)
+	b.SetState(StateProvision)
 
 	return nil
 }
 
-func (builder *Builder) HandleProvision(w http.ResponseWriter, r *http.Request) (err error) {
-	builder.GuardState(StateProvision)
-
-	if r.Method != "PUT" {
-		return fmt.Errorf("Builder.HandleProvision: Unexpected request method")
-	}
-
+func (b *Builder) HandleProvision(w http.ResponseWriter, r *http.Request) error {
 	logrus.WithFields(logrus.Fields{"argBuildPath": argBuildPath}).Debug("Builder.HandleProvision: Opening manifest.json")
 
 	dst, err := os.OpenFile(
```

Comment:
I think it's fine to just use `os.Create()` here, AFAICT there are no security benefits from creating non-writable files or doing O_EXCL, the inputs come from trusted sources and if someone owns the worker-worker site 0400 will not really stop any attackers. Or am I missing something here?

Reply:
It wasn't much meant as a security measure but more as an assertion that the machine the builder is running on is in a correct state (no output has been written yet).

However since this makes running the builder locally more annoying and the gain is small I've switched this to `os.Create`.